File: problem_21/amicable_numbers_test.rb
require 'minitest/test'
require 'minitest/autorun'

class AmicableNumbersTest < Minitest::Test
  def amicable_number_sum(max_num)
    amicables = []

    (1..max_num).each do |a|
      b = divisor_sum(a)
      amicables.push(a, b) if a != b && a == divisor_sum(b)
    end

    amicables.uniq.sum
  end

  def divisor_sum(num)
    (1..(num/2)).find_all { |i| num % i == 0 }.sum
  end

  #----------------- Tests -----------------#

  def test_amicable_number_sum
    assert_equal 504,   amicable_number_sum(1000)
    assert_equal 31626, amicable_number_sum(10000)
  end

  def test_divisor_sum
    assert_equal 284, divisor_sum(220)
    assert_equal 220, divisor_sum(284)
  end
end

File: problem_21/amicable-numbers.test.js
function amicableNumberSum(maxNum) {
  let sum = 0;

  for (let a = 1; a < maxNum; a++) {
    let b = divisorSum(a);
    if (a !== b && divisorSum(b) === a) {
      sum += a;
    }
  }

  return sum;
}

function divisorSum(num) {
  let sum = 0;

  for (let i = 1; i <= num / 2; i++) {
    if (num % i === 0) {
      sum += i;
    }
  }

  return sum;
}

//----------------- Tests -----------------//

test("amicableNumberSum", () => {
  expect(amicableNumberSum(1000)).toBe(504);
  expect(amicableNumberSum(10000)).toBe(31626);
});

test("divisorSum", () => {
  expect(divisorSum(220)).toBe(284);
  expect(divisorSum(284)).toBe(220);
});
